### ISSUE TITLE
Request: make itch-app available on flathub

### ISSUE THREAD

**Issue by fabiscafe** [itchio/itch#1624]
It would be so nice to have itch available for all linux distros, by building a [flatpak](https://flatpak.org/) and [submit it to flathub](https://github.com/flathub/flathub/wiki).

This way it would show up in gnome-software and KDE-discover for every user. Please make this possible :+1:

**Comment by fasterthanlime** [comment]
Ah, the trinity is complete! An AppImage and Snap have already been requested - but none of those are planned for the near future.

Currently, we build:

  * A portable build (all distros)
  * An .rpm package
  * A .deb package
  * An ArchLinux PKGBUILD

I really want to get rid of the .rpm/.deb (or at least, have them not contain the full app), see https://github.com/itchio/itch/issues/827

As for AppImage/Snap/flatpak, I haven't made up my mind yet - but I'd really like the itch app to leverage our [own patching technology](http://github.com/itchio/wharf).

Anyway, it's good to have a canonical issue for this, so thanks for opening it! 🌟

**Comment by fabiscafe** [comment]
I don't know how you handle this right now. Do you install the final binary into the users home directory, so an app- update would be possible without the need of root permissions?

If this is the case, it would be the same as for steam in flatpak. Otherwise it would make things a bit harder. But flatpak provides Delta/Diff updates. So even then updates would be minimal.

For snap, I'm not a fan of, because it's bound to Ubuntu base, don't provides a working sandbox on systems without apparmor (everything except Ubuntu) and is just again a project of canonical that is not well made... Like Mir/Unity, upstart, Ubuntu one, etc.

Appimages would be great for the first step, but updating them is also a pain. They also don't integrate into any software store. 

Just to help you to decide 😁. And yes, I don't think Deb,rpm,AUR are the right thing to do.(said as a Manjaro/Archlinux user).

**Comment by fasterthanlime** [comment]
> I don't know how you handle this right now. Do you install the final binary into the users home directory, so an app- update would be possible without the need of root permissions?

It's not done yet, but that's basically the plan, yep!

I don't doubt that by now both Flatpak & Snap have some kind of diff updates, but my reservations go something like:

  * Store approval is a pain (especially if every update needs to be manually reviewed)
  * I welcome sandboxes, but have had a rather negative experience with Snap's one - the app needs to access a bunch of things to install games properly (sometimes even exec `gksudo` to install system packages, with the user's approval), and sandboxes tend not to like that. Note that [the itch app does its own sandboxing](https://itch.io/docs/itch/using/sandbox/linux.html)
  * The more types of packages we generate, the longer it takes us to push an update (and we want to be able to push updates in minutes - something our [uploading tool](http://itch.io/docs/butler) handles beautifully) - that's one of the reasons I want to get rid of RPM/DEB/PKGBUILD in the first place :)
  * Flatpak/Snap adoption is an issue - there's no clear winner yet - if they're the only option, it'll turn off some users who "just want the darn app", not another package manager, possibly some runtimes, etc.
  * Having [our own setup system](https://github.com/itchio/itchSetup) lets us install & update the app using the same system on all 3 platforms (Linux, macOS & Windows) - Flatpak/Snap is another system we need to potentially look into if things break.

---

Ideally, I'd like for the app to be

  1. Easily installed on all 3 platforms (preferably with an installer that downloads the latest from itch.io directly, extracts on the fly, like itchSetup already does)
  2. Updated in the same way on all 3 platforms, which is:
    a. The app transparently looks for new versions in the background
    b. The app patches itself to a new folder in the background
    c. A notification is shown with the changelog and the option to restart the app now, or wait until later (that's what the macOS & Windows versions of the app do, although they're based on [Squirrel](https://github.com/squirrel) for now - which I have other gripes with).

Maybe itchSetup will be packaged as many different types of package, like RPM/DEB/PKGBUILD/Snap/Flatpak/whichever - but the actual app will probably be patched with our own stuff, just because it's immensely more convenient for us. (Patching has been a big part of my work for the past 1.5 years, we've got a really good thing going).

**Comment by fabiscafe** [comment]
> Easily installed on all 3 platforms (preferably with an installer that downloads the latest from itch.io directly, extracts on the fly, like itchSetup already does)

Thats nothing to make yourself friends on linux. I'm around for 12 years on linux and started with some debian and later Arch packaging. But it's always the same thing. Generic installer "Windows like things" will just not reach the majority because nobody wants them in it's distro. Look back at the time when [MojoSetup](https://icculus.org/mojosetup/) and [Loki installer](https://github.com/megastep/loki_setup) was a thing. Both died because, on Linux, nobody wants external updater/installer..

> Updated in the same way on all 3 platforms, which is: [...]

That seems to be not ideal, but could work out, since Itch is its own "appstore".

one little thing, befor I go on. 

> b. The app patches itself to a new folder in the background

you should always use XDG-standard folders. https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html
Just like in this case it would be $XDG_DATA_HOME, wich (if not set) is /home/$USER/.local/share/ Thats more and more important for the Linux desktop. Just because i've seen that you install games right now to ~/.config. Thats the completely wrog place. :+1: 

but BTT

* Store approval is not a thing in flatpak. You have this great flathub, but if you dont like it - you can just host your own flatpak repo. So no waiting required at all. (But I have to say that the flathub people are very nice. So I dont see a reaso for now to not make use of this one.

* Sandboxing can completely turned off in flatpaks (or just parts of it)
* flatpak can ship the full runtime bundled. So no external requirements. There are only 2 Moments when you really need root permissions: to install the flatpak system and to install an additional flatpak repo. Everything else can be done (and default to) "user-only-mode"

* builds itself. I think you would have to split this up. Install the Itch "appstore" as an repo package via some kind of packagemanager, but install the games itself as "extensions" to Itch. Thats maybe the best Idea to have common path on all systems. I also dont think People would try to uninstall dokidokiliteratureclub via Gnome Store, if installed via itch. So thats OK.

* flatpak vs snap. From what I can tell, you will not get a winner. I think snap is what will "win" on the Ubuntu Server side. Because nobody expect them to be secured. But they are fast any easy to update without a big dependency check. Just with the command line. Also companies can shipp preset build, like Owncloud does. So no need for a user to setupo its own apache. On the Desktop side, you have only the ubuntu version of the gnome software "store" that supports snap. Nobody else does.
Flatpak on the other side is supported by Gnome and by KDE with gnome software and discover(in its next version). Its also supported by redhat and the community itself. Thats something I really like about it. There is also this blog entry by the solus devs: 
https://solus-project.com/2017/01/18/adopting-flatpak-to-reassemble-third-party-applications/ 

* Having your own setup system will end with people who repackage Itch as PPA, AUR and at some point flatpak. Just look at the situation with Telegram. They ship an binary with auto updater and auto installer and whats the point at the end:
https://packages.debian.org/search?suite=sid&searchon=names&keywords=telegram
https://launchpad.net/~atareao/+archive/ubuntu/telegram
https://www.archlinux.org/packages/community/x86_64/telegram-desktop/
https://snapcraft.io/telegram-latest/
https://flathub.org/repo/appstream/org.telegram.desktop.flatpakref

None of them are official supported. Some of them are outdated. Its just what's the Linux way of doing things. ( And I hate this :( )

**Comment by sigh-snapvsflatpakagain** [comment]
>On the Desktop side, you have only the ubuntu version of the gnome software "store" that supports snap. Nobody else does. Flatpak on the other side is supported by Gnome and by KDE with gnome software and discover(in its next version).

Simply Not True. Discover that you mention will soon not only have the Flatpak support, [but also Snap](https://pointieststick.wordpress.com/2018/01/16/this-week-in-discover/), the user will be able to chose whatever they like better. Listings will be de-duplicated so that apps supporting multiple formats are only listed once based on the users preference. Solus that you also mention got [a updated view on Snap](https://www.youtube.com/watch?v=3KNKtZNfAHE) and they are also building the [linux-steam-integration (LSI)](https://github.com/solus-project/linux-steam-integration) with Snap because they figured that for them "[it is dramatically simpler to do this using Snaps](https://plus.google.com/+Solus-Project/posts/ZL8C2wBqbfg)". Confinement can be adjusted, delta updates supported, and Snaps don't require lengthy store approval they either (apps can be pushed out almost instantly through automated review). Etc...

This so typical war over the technologies is silly. I don't see the companies Canonical and Red Hat fighting over Snap vs Flatpak (actually their cooperation is healthier than ever with Ubuntu now back on Gnome), but a few vocal individuals spreading FUD that just makes Linux in general look like a childish battlefield instead of the joyful arena we know and love. The two technologies Flatpak and Snap are both great and there's no reason to fight one over the other, Discover is showing in practice that they both can live side by side. FUD is really not helpful for any developer wanting to support Linux, so please stop doing that.

If you're actually interested to learn about Snap I suggest that you read the [documentation](https://snapcraft.io/), watch some [videos](https://www.youtube.com/channel/UCcH6oAZ0FOSVUMUAojHtFBg/videos), get help from [Snapcrafters](https://github.com/snapcrafters), ask on the [forum](https://forum.snapcraft.io) or in [Rocket Chat](https://rocket.ubuntu.com). There's really no excuse to keep repeating old or false information. Same thing to be said about individuals spreading FUD about Flatpak, please stop doing that. Thanks.

**Comment by fasterthanlime** [comment]
@sigh-snapvsflatpakagain rest assured that I'm staying neutral in all this - I'm firmly convinced that all package systems, on all OSes, and including our own solutions - suck in equally annoying ways ;)

I'll take a deeper look at providing both once v25 is ready for release. For now, there's too much development left for me to worry a lot about packaging.

**Comment by ghost** [comment]
FYI:

Snap is made by Canonical, keep in mind what happened with everything else they created - Upstart lost to Systemd, Mir lost to Wayland, Unity lost to GNOME... Snaps will lose to Flatpak, simple as that. Also if you are not sure which tech is going to prevail in Linux desktop ecosystem, just check which one Red Hat is support and you are good to go - whatever RH pushes, it stays, that's the way it is like it or not.

Flatpak is the future, invest (also I personally refuse to run Itch.io till it gets proper sandboxing, which flatpak provides, I don't trust Firefjail which had like gazyllion vulnerabilities discovered last year).

**Comment by quyse** [comment]
> Flatpak is the future, invest (also I personally refuse to run Itch.io till it gets proper sandboxing, which flatpak provides, I don't trust Firefjail which had like gazyllion vulnerabilities discovered last year).

@gutigen Is it actually possible to use Flatpak as a sandbox for non-flatpaked games from itch.io? And if we are talking about accidentally running deliberate malware, will it help at all? Limiting filesystem access is good, but if game needs access to DRI/audio (basically every game with graphics and sound does) it gets very big attack surface anyway.

**Comment by ghost** [comment]
@quyse Flatpaked application sees only parent bubblewrap process and nothing else, however most apps still need access to IPC, network, devices. It for sure limits attack surface, but doesn't isolate completely, so devs are working on Pipewire, Wayland and Flatpak combo that should solve this, but Pipewire is still in early stages, so it won't happen anytime soon (Pipewire is replacement for Pulseaudio/Jack made with low latency audio and security in mind, it's slowly getting integrated into Fedora).

**Comment by fasterthanlime** [comment]
merging with #2074

**Comment by MasterionX** [comment]
> FYI:
> 
> Snap is made by Canonical, keep in mind what happened with everything else they created - Upstart lost to Systemd, Mir lost to Wayland, Unity lost to GNOME... Snaps will lose to Flatpak, simple as that. Also if you are not sure which tech is going to prevail in Linux desktop ecosystem, just check which one Red Hat is support and you are good to go - whatever RH pushes, it stays, that's the way it is like it or not.
> 
> Flatpak is the future, invest (also I personally refuse to run Itch.io till it gets proper sandboxing, which flatpak provides, I don't trust Firefjail which had like gazyllion vulnerabilities discovered last year).

lolno

[Snaps](https://uappexplorer.com/snaps): >2500
[Flatpak apps](https://flathub.org/apps/category/All): >600

Red Hat is just not desktop-oriented, simple as that.
